[ADD] odoo-bin cloc cli subcommand. (#52016)
odoo/odoo: doc/reference/cmdline.rst

    $ odoo_bin scaffold my_module /addons/

This will create module *my_module* in directory */addons/*.

Cloc
====

.. program:: odoo-bin cloc

Odoo Cloc is a tool to count the number of relevant lines written in
python, javascript or xml. This can be used as a rough metric for pricing
maintenance of customizations.

It has two modes of operation, either by providing a path:

.. option:: --path PATH, -p PATH

Or by providing the name of a database:

.. option:: --database DATABASE, -d DATABASE

The addons path should be provided as well

.. code-block:: console

    $ odoo-bin cloc --addons-path=dirs -d database

In the latter mode, only the custom code is accounted for.
You can also use a configuration file -c configuration.conf instead of the addons-path

.. code-block:: console

    $ odoo-bin cloc -c configuration.conf -d database

both mode can be used simultaneously to count the line for a given folder
in addition of the lines in the given database.

.. code-block:: console

    $ odoo-bin cloc --addons-path=dirs -d database -p module_path

The verbose option show the details per files.

.. option:: --verbose, -v

> Which files Odoo Cloc counts ?
> ------------------------------

Reviewer: ```suggestion
Processed files
---------------
```

> Database mode
> '''''''''''''

Reviewer: ```suggestion
.. _reference/cmdline/cloc/database-option:

With the :option:`--database` option
''''''''''''''''''''''''''''''''''''
```
To avoid implying that it excludes using `--path`

> Odoo Cloc counts the lines in each file of non standard installed modules in a
> given database. In addition, it counts the python line from server action and
> custom computed field stored directly in the database, in other words, not defined
> in any module or defined in an importable module.

Reviewer: ```suggestion
Odoo Cloc counts the lines in each file of third party installed modules in a
given database. In addition, it counts the python lines of server actions and
custom computed fields stored directly in the database. In other words, that are
not defined in any standard or importable module.
```

> Some files are excluded from the count by default
> 
> - The manifest (__manifest__.py or __openerp__.py)
> - The content of the folder static/lib
> - The tests defined in folder tests and static/tests
> - The xml files declared in demo or demo_xml section of the manifest

Reviewer: ```suggestion
Some files are excluded from the count by default:

- The manifest (:file:`__manifest__.py` or :file:`__openerp__.py`)
- The content of the folder :file:`static/lib`
- The tests defined in the folder :file:`tests` and :file:`static/tests`
- The xml files declared in the ``demo`` or ``demo_xml`` sections of the manifest
```

> Finally, it's possible to define in each module a list of files to ignore
> with the directive "cloc_exclude" in the manifest.

Reviewer: ```suggestion
A list of files that should be ignored by Odoo Cloc can be defined per module. This
is specified by the ``cloc_exclude`` entry of the manifest:
```

.. code-block:: python

    "cloc_exclude": [
        "lib/common.py", # exclude a single file
        "data/*.xml", # exclude all xml file in a specific folder
        "example/**/*", # exclude all file in the folder and its subfolder
    ]

> The syntax "\*\*/\*" can be used to ignore a complete module. In that case,
> the module will not be covered by the maintenance service.
> 
> For more information about the syntax,
> see `glob <https://docs.python.org/3.6/library/pathlib.html#pathlib.Path.glob>`_

Reviewer: ```suggestion
| The syntax ``\*\*/\*`` can be used to ignore an entire module. This can be useful
  to exclude a module from maintenance service costs.
| For more information about the syntax, see `glob
  <https://docs.python.org/3.6/library/pathlib.html#pathlib.Path.glob>`_.
```
We will use the tool for that purpose but it's not the standard.

> Path mode
> '''''''''

Reviewer: ```suggestion
.. _reference/cmdline/cloc/path-option:

With the :option:`--path` option
''''''''''''''''''''''''''''''''
```

> Odoo Cloc counts as described in database mode if a manifest file is present in
> the given folder. Otherwise, it counts all files.

Reviewer: ```suggestion
The method works the same as with :ref:`the --database option
<reference/cmdline/cloc/database-option>` if a manifest file is present in the given
folder. Otherwise, it counts all files.
```

> How does cloc make the difference between standard and custom module ?
> ----------------------------------------------------------------------
> 
> Odoo Cloc uses the following heuristic: If the module is in the same addons path as
> the modules base, web or web_enterprise then the module is considered as standard.
> 
> Therefore, we strongly advise to have your custom module in a separate addons_folder

Reviewer: ```suggestion
Identifying third party modules
-------------------------------

To make the difference between standard and third party modules, Odoo Cloc uses the
following heuristic: If a module is in the same addons path as the modules
``base``, ``web`` or ``web_enterprise`` then it is considered as standard.
Therefore, we strongly advise to place your custom modules in a separate folder.
```

Error Handling
--------------

Some file cannot be counted by Odoo Cloc.
Those file are reported at the end of the output.

Max file size exceeded
''''''''''''''''''''''

> Odoo Cloc rejects any file bigger than 25MB. Usually source files are smaller
> than 1 MB, if you have that kind of file reported, it's probably
> - a generated xml file that contains lot of data, exclude it in the manifest
> - a javascript library that should be placed in static/lib

Reviewer: ```suggestion
Odoo Cloc rejects any file larger than 25MB. Usually, source files are smaller
than 1 MB. If a file is rejected, it may be:

- A generated xml file that contains lots of data. Exclude it in the manifest.
- A JavaScript library that should be placed in the :file:`static/lib` folder.
```

Syntax Error
''''''''''''

> Odoo Cloc cannot count the lines of code of a python file with a wrong syntax. If
> your module contains such files, either the file should be fixed otherwise the
> module will not work. If the module works with those files, they are not used and
> therefore should be removed from the module or at least excluded from the count.

Reviewer: ```suggestion
Odoo Cloc cannot count the lines of code of a python file with an incorrect syntax.
If your module contains such files, they should be fixed to allow the module to
work. If the module works despite the presence of those files, this is because they
are not used and they should therefore be removed from the module, or at least
excluded from Odoo Cloc's count.
```

Odoo Cloc counts the lines in each file of extra installed modules in a
given database. In addition, it counts the python lines of server actions and
custom computed fields stored directly in the database. In other words, that are
not defined in any standard or importable module.

Some files are excluded from the count by default:

- The manifest (:file:`__manifest__.py` or :file:`__openerp__.py`)
- The content of the folder :file:`static/lib`
- The tests defined in the folder :file:`tests` and :file:`static/tests`
- The xml files declared in the ``demo`` or ``demo_xml`` sections of the manifest

A list of files that should be ignored by Odoo Cloc can be defined per module. This
is specified by the ``cloc_exclude`` entry of the manifest:

.. code-block:: python

    "cloc_exclude": [
        "lib/common.py", # exclude a single file
        "data/*.xml", # exclude all xml file in a specific folder
        "example/**/*", # exclude all file in the folder and its subfolder
    ]

| The syntax ``**/*`` can be used to ignore an entire module. This can be useful
  to exclude a module from maintenance service costs.
| For more information about the syntax, see `glob

>   <https://docs.python.org/3.6/library/pathlib.html#pathlib.Path.glob>`_.

Reviewer 2: Perhaps don't pin the version, could even remove it entirely, docs.python.org redirects to the latest one automatically.

```suggestion
  <https://docs.python.org/3/library/pathlib.html#pathlib.Path.glob>`_.
```
